### PR TITLE
clean up the balancing code around the batched vs deprecated way of sampling segments to balance

### PR DESCRIPTION
**Pull request by capistrant**
**Start release notes suggestion**

The dynamic coordinator config, `percentOfSegmentsToConsiderPerMove` has been deprecated and will be removed in a future release of Druid. It is being replaced by a new segment picking strategy introduced in https://github.com/apache/druid/pull/11257. This new strategy is currently toggled off by default, but can be toggled on if you set the dynamic coordinator config `useBatchedSegmentSampler` to `true`. Setting this as such, will disable the use of the deprecated `percentOfSegmentsToConsiderPerMove`. In a future release, `useBatchedSegmentSampler` will become permanently true.

**End release notes suggestion**

#### Fixed unintended code path execution in BalancerStrategy#pickSegmentsToMove

https://github.com/apache/druid/pull/11257 Introduced a new way of sampling segments for balancing that is much more performant than the legacy way that exists today. There is a case in the code where the deprecated code can be executed despite, the dynamic config for batched segment sampling being enabled. In [this](https://github.com/yuanlihan/druid/blob/b02a9e659846a88c7be375ffb805340c38d1dbf5/server/src/main/java/org/apache/druid/server/coordinator/duty/BalanceSegments.java#L198) line of code, if `useBatchedSegmentSampler() == true` and `maxSegmentsToMove == 1`, you would expect the batched segment sampler code to execute. However, you can see [here](https://github.com/yuanlihan/druid/blob/b02a9e659846a88c7be375ffb805340c38d1dbf5/server/src/main/java/org/apache/druid/server/coordinator/BalancerStrategy.java#L82) that the new batched sampler is not used in this case because reservoir size is 1. This PR fixes this behavior by splitting `BalancerStrategy#pickSegmentsToMove(List<ServerHolder>, Set<String>, int, double)` into two methods using overloading - one for the new batched sampler, and one for the deprecated way of sampling one segment at a time. This change will also simplify cleanup of the deprecated code in what I assume will be the 0.24.X cycle.

#### Cleaned up the code for the competing segment pick strategies (deprecated vs batched segment sampler)

I did a few things to make the code more understandable and to make it abundantly clear that the old way of sampling segments to move with the help of the dynamic config `percentOfSegmentsToConsiderPerMove` is deprecated:
* Added `@Deprecated` annotations to the `CoordinatorDynamicConfig` code related to `percentOfSegmentsToConsiderPerMove`
* Updated the Coordinator Dynamic Config docs to explicitly state that `percentOfSegmentsToConsiderPerMove` is deprecated and will have no effect if `useBatchedSegmentSampler` dynamic config is `true`
* Split `BalancerStrategy#pickSegmentsToMove` into two methods with the same name and overloaded arguments. One implementation is deprecated and takes a double for `percentOfSegmentsToConsiderPerMove` while the other is the new batched sampler which takes an `int` for the reservoir size.
  * This change also makes `BalanceSegments#balanceServers` more understandable to those with limited familiarity with this work to migrate to the batched segment sampler. Before the call to `BalancerStrategy#pickSegmentsToMove` was opaque because it included both the reservoir size and the percent of segments to consider per move despite only one of those arguments being used depending on the value of the reservoir size (this also relates to the unintended code execution I talk about above). The new way makes it clear that if batched segment sampling is enabled, the non-deprecated `pickSegmentsToMove` is called. And if it is not, the deprecated `pickSegmentsToMove` is called. 

<hr>

##### Key changed/added classes in this PR
 * `BalanceSegments`
 * `CoordinatorDynamicConfig`
 * `BalancerStrategy`

<hr>

This PR has:
- [x] been self-reviewed.
- [X] added documentation for new or modified features or behaviors.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by capistrant** [comment]
@yuanlihan could you take a look at this since you wrote the original code for the batched sampler! I hope my code makes sense, but please comment on any mistakes I may have made!

File: server/src/main/java/org/apache/druid/server/coordinator/duty/BalanceSegments.java [hunk 1]
```diff
@@ -202,12 +202,20 @@ private void balanceTier(
     final int maxToLoad = params.getCoordinatorDynamicConfig().getMaxSegmentsInNodeLoadingQueue();
     int moved = 0, unmoved = 0;
 
-    Iterator<BalancerSegmentHolder> segmentsToMove = strategy.pickSegmentsToMove(
-        toMoveFrom,
-        params.getBroadcastDatasources(),
-        params.getCoordinatorDynamicConfig().useBatchedSegmentSampler() ? maxSegmentsToMove : DEFAULT_RESERVOIR_SIZE,
```
Comment (yuanlihan) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/BalanceSegments.java:208]: I think we can delete this unused constant variable

**Comment by yuanlihan** [comment]
> @yuanlihan could you take a look at this since you wrote the original code for the batched sampler! I hope my code makes sense, but please comment on any mistakes I may have made!

Hi @capistrant
It looks good to me. Thanks for making the code neat!

File: docs/configuration/index.md [hunk 2]
```diff
@@ -886,7 +886,7 @@ Issuing a GET request at the same URL will return the spec that is currently in
...
+|`percentOfSegmentsToConsiderPerMove`|Deprecated. Will eventually be phased out by the batched segment sampler. Only used if `useBatchedSegmentSampler == false`. The percentage of the total number of segments in the cluster that are considered every time a segment needs to be selected for a move. Druid orders servers by available capacity ascending (the least available capacity first) and then iterates over the servers. For each server, Druid iterates over the segments on the server, considering them for moving. The default config of 100% means that every segment on every server is a candidate to be moved. This should make sense for most small to medium-sized clusters. However, an admin may find it preferable to drop this value lower if they don't think that it is worthwhile to consider every single segment in the cluster each time it is looking for a segment to move.|100|
```
Comment (a2l007) [inline docs/configuration/index.md:889]: Should we add a similar deprecation message in the web console description for this property?

**Comment by a2l007** [comment]
Tagging this under Release Notes so that users are aware of the deprecation of `percentOfSegmentsToConsiderPerMove`

File: docs/configuration/index.md (hunk 2 above)
Comment (capistrant) [reply docs/configuration/index.md:889]: yep, good call @a2l007

**Review by a2l007 — APPROVED** [review]

**Comment by capistrant** [comment]
@a2l007 @yuanlihan thank you both for reviewing! I synced up with master and plan on merging once CI is green.

File: docs/configuration/index.md (hunk 2 above)
Comment (jihoonson) [reply docs/configuration/index.md:889]: Can we document what the preferred alternative is here?

File: web-console/src/druid-models/coordinator-dynamic-config.tsx [hunk 3]
```diff
@@ -213,14 +213,16 @@ export const COORDINATOR_DYNAMIC_CONFIG_FIELDS: Field<CoordinatorDynamicConfig>[
...
       <>
-        The percentage of the total number of segments in the cluster that are considered every time
-        a segment needs to be selected for a move. Druid orders servers by available capacity
-        ascending (the least available capacity first) and then iterates over the servers. For each
-        server, Druid iterates over the segments on the server, considering them for moving. The
-        default config of 100% means that every segment on every server is a candidate to be moved.
-        This should make sense for most small to medium-sized clusters. However, an admin may find
-        it preferable to drop this value lower if they don&apos;t think that it is worthwhile to
-        consider every single segment in the cluster each time it is looking for a segment to move.
+        Deprecated. This will be phased out by the batched segment sampler. Only used if
```
Comment (jihoonson) [inline web-console/src/druid-models/coordinator-dynamic-config.tsx:216]: Can we document what the preferred alternative is here as well?

**Review by jihoonson — APPROVED** [review]
LGTM. Please consider my comments before merging it.

File: web-console/src/druid-models/coordinator-dynamic-config.tsx (hunk 3 above)
Comment (capistrant) [reply web-console/src/druid-models/coordinator-dynamic-config.tsx:216]: could you clarify what you mean by this? I take it as you asking me to state that the preferred alternative is to set `useBatchedSegmentSampler == true`. so perhaps?

> Deprecated. Will eventually be phased out by the batched segment sampler. You can enable the batched segment sampler now by setting the dynamic Coordinator config, `useBatchedSegmentSampler`, to `true`. Note that if you choose enable the batched segment sampler, this config has no effect on balancing. If `useBatchedSegmentSampler == false`, this config defines the percentage of the total number of segments in the cluster that are considered every time a segment needs to be selected for a move. Druid orders servers by available capacity ascending (the least available capacity first) and then iterates over the servers. For each server, Druid iterates over the segments on the server, considering them for moving. The default config of 100% means that every segment on every server is a candidate to be moved. This should make sense for most small to medium-sized clusters. However, an admin may find it preferable to drop this value lower if they don't think that it is worthwhile to consider every single segment in the cluster each time it is looking for a segment to move.|100|

Comment (jihoonson) [reply web-console/src/druid-models/coordinator-dynamic-config.tsx:216]: Sorry for ambiguity. Yes, encouraging using `useBatchedSegmentSampler` instead will be enough. The one in your comment looks good to me.

**Comment by jihoonson** [comment]
@capistrant thank you for updating docs. Feel free to merge once the "docs" CI passes.

**Comment by jihoonson** [comment]
The "docs' CI passed. Merging this PR.

**Pull request merged**